Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+		 * If some callback is not provided (NULL) the request will always succeeds.
+		 */
+		struct KniOldIoctlCallbacks
+		{
+			/**
+			 * Pointer to function of changing MTU.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*change_mtu)(uint8_t port_id, unsigned int new_mtu);
+			/**
+			 * Pointer to function of configuring network interface.
+			 * Must return 0 in case of success or negative error code
+			 */
+			int (*config_network_if)(uint8_t port_id, uint8_t if_up);
+		};
+
+		/**
+		 * @brief KNI device initialization data.
+		 * Used to create new KNI device.
+		 * Usage of callbacks member or oldCallbacks member is defined by
+		 * result of KniDevice#callbackVersion
+		 */
+		struct KniDeviceConfiguration
+		{
+			enum
+			{
+				// Must correspond to minimum of
+				// RTE_KNI_NAMESIZE [DPDK] = 32 and
+				// IFNAMSIZ [LINUX] = 16 (on most systems)
+				KNI_NAME_SIZE = 16
+			};
+			/**
+			 * Name used to display device in system.
+			 * Must not interfere with already existing network interfaces.
+			 */
+			char name[KNI_NAME_SIZE];
+			union
+			{
+				KniIoctlCallbacks* callbacks;
+				KniOldIoctlCallbacks* oldCallbacks;
+			};
+			/**
+			 * Pointer to MAC (ETHERNET) address of new KNI device.
+			 * If omitted (NULL) some valid address will be automatically generated.
+			 * If provided will be cached by new KNI device info structure.
+			 */
+			MacAddress* mac;
+			/**
+			 * Used in same way as DPDK port id.
+			 * If some new callbacks are omitted then have separate meaning
+			 * (see KniIoctlCallbacks#config_mac_address and KniIoctlCallbacks#config_promiscusity).
+			 * Can be set to UINT16_MAX.
+			 */
+			uint16_t portId;
+			/** MTU of new KNI device. Will be cached by new KNI device info structure */
+			uint16_t mtu;
+			/**
+			 * If set forces to bind KNI Linux kernel thread (NOT userspace thread) to specific core.
+			 * If rte_kni kernel module is loaded with "kthread_mode=single" then -
+			 * rebinds kernel thread used for all KNI devices to specified core.
+			 * If rte_kni kernel module is loaded with "kthread_mode=multiple" then -
+			 * binds new kernel thread for this device to specified core.
+			 */
+			bool bindKthread;
+			/** ID of core to bind Linux kernel thread to (same as DPDK cores IDs) */
+			uint32_t kthreadCoreId;
+		};
+
+	private:
+		/** All instances of this class MUST be produced by DeviceFactory factory */
+		KniDevice(const KniDeviceConfiguration& conf, size_t mempoolSize, int unique);
+		/** This class is not copyable */
+		KniDevice(const KniDevice&);
+		/** This class is not copyable */
+		KniDevice& operator=(const KniDevice&);
+
+	protected:
+		friend struct KniDeviceList;
+		/**
+		 * All instances of this class MUST be destroyed by DestroyDevice method
+		 * or automatically on application exit by KniDeviceList singleton.
+		 */
+		~KniDevice();
+
+	public:
+		/* Device manipulation */
```

Comment:
For other devices like DpdkDevice the device list not only handels the creation and storage of devices, but some other useful functions. For KNI this is not the case - only creation/search/storage/deletion no more. So this is no need for user to learn the documentation of two classes, the relations between them and search in two files.
The static methods with proper name will have self-documeted property.
The `callbackVersion`, `isCallbackSupported` are only exist because i don't now any better way to express version eternal API dependency in C++ especially that is fixed in compile time

Reply:
People who are using PcapPlusPlus are already used to this convention: one class that represents the device itself and another one that represents the device list. It is like that everywhere: 
- `PcapLiveDevice`/`WinPcapLiveDevice` and `PcapLiveDeviceList`
- `PcapRemoteDevice` and `PcapRemoteDeviceList`
- `DpdkDevice` and `DpdkDeviceList`
- `PfRingDevice` and `PfRingDeviceList`

I think you should align `KniDevice` to the same convention to avoid confusion and keep uniformity among all devices

Reply:
Ok. Do You suggest to create separate h/cpp pair for that class?

Reply:
Yes, I think that would help

Reply:
\+